File: pkg-r/tests/testthat/test-network.R
test_that("network() builds an htmlwidget with the expected payload", {
  nodes <- data.frame(
    id = c("a", "b", "c"),
    x = c(0, 1, 2),
    y = c(0, 1, 0),
    size = c(4, 8, 6),
    group = c("G1", "G1", "G2"),
    label = c("Alpha", "Beta", "Gamma"),
    stringsAsFactors = FALSE
  )
  edges <- data.frame(
    source = c("a", "b"),
    target = c("b", "c"),
    weight = c(1.5, 2),
    stringsAsFactors = FALSE
  )
  w <- network(nodes, edges,
    layout = "precomputed", iterations = 100, label_threshold = 5
  )

  expect_s3_class(w, "htmlwidget")
  expect_equal(w$x$data$columns$id, c("a", "b", "c"))
  expect_equal(w$x$data$columns$source, c("a", "b"))
  expect_equal(w$x$data$columns$target, c("b", "c"))
  expect_equal(w$x$data$columns$x, c(0, 1, 2))
  expect_equal(w$x$data$columns$size, c(4, 8, 6))
  expect_equal(w$x$data$columns$weight, c(1.5, 2))
  expect_equal(w$x$data$meta$nodeGroup, c("G1", "G1", "G2"))
  expect_equal(w$x$data$meta$nodeLabels, c("Alpha", "Beta", "Gamma"))
  expect_equal(w$x$options$layout, "precomputed")
  expect_equal(w$x$options$iterations, 100)
  expect_equal(w$x$options$labelThreshold, 5)
})

test_that("network() omits optional columns when absent", {
  nodes <- data.frame(id = c("a", "b"), stringsAsFactors = FALSE)
  edges <- data.frame(
    source = "a", target = "b", stringsAsFactors = FALSE
  )
  w <- network(nodes, edges)

  expect_null(w$x$data$columns$x)
  expect_null(w$x$data$columns$size)
  expect_null(w$x$data$columns$weight)
  expect_null(w$x$data$meta$nodeGroup)
  expect_null(w$x$data$meta$nodeLabels)
  expect_equal(w$x$options$layout, "forceatlas2")
})

test_that("network() forwards a palette override", {
  nodes <- data.frame(id = c("a", "b"), stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "b", stringsAsFactors = FALSE)
  w <- network(nodes, edges, palette = c("#111111", "#222222"))
  expect_equal(w$x$options$palette, c("#111111", "#222222"))
})

test_that("network() validates its input", {
  nodes <- data.frame(id = "a", stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "a", stringsAsFactors = FALSE)

  expect_error(network(list(id = "a"), edges), "must be a data frame")
  expect_error(network(nodes, list(source = "a")), "must be a data frame")
  expect_error(
    network(data.frame(foo = 1), edges), "must contain an `id`"
  )
  expect_error(
    network(nodes, data.frame(a = 1, b = 2)), "must contain `source` and `target`"
  )
})

test_that("network() rejects duplicate node ids", {
  nodes <- data.frame(id = c("a", "a"), stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "a", stringsAsFactors = FALSE)
  expect_error(network(nodes, edges), "duplicate node id\\(s\\): a")
})

test_that("network() rejects an edge referencing a missing node", {
  nodes <- data.frame(id = c("a", "b"), stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "z", stringsAsFactors = FALSE)
  expect_error(
    network(nodes, edges),
    "edge endpoint\\(s\\) not found among node ids: z"
  )
})

test_that("network() requires x/y when layout is precomputed", {
  nodes <- data.frame(id = c("a", "b"), stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "b", stringsAsFactors = FALSE)
  expect_error(
    network(nodes, edges, layout = "precomputed"),
    "requires `x` and `y`"
  )
})

test_that("network() rejects a factor coordinate column", {
  nodes <- data.frame(
    id = c("a", "b"), x = factor(c("0", "1")), y = c(0, 1),
    stringsAsFactors = FALSE
  )
  edges <- data.frame(source = "a", target = "b", stringsAsFactors = FALSE)
  expect_error(network(nodes, edges), "`x` must be numeric")
})

test_that("network() rejects empty nodes", {
  nodes <- data.frame(id = character(0), stringsAsFactors = FALSE)
  edges <- data.frame(
    source = character(0), target = character(0), stringsAsFactors = FALSE
  )
  expect_error(network(nodes, edges), "`nodes` has no rows/cells")
})

test_that("network() forwards a theme override", {
  nodes <- data.frame(id = c("a", "b"), stringsAsFactors = FALSE)
  edges <- data.frame(source = "a", target = "b", stringsAsFactors = FALSE)
  w <- network(nodes, edges, theme = list(background = "#333"))
  expect_equal(w$x$options$theme$background, "#333")
  expect_null(network(nodes, edges)$x$options$theme)
})

File: pkg-py/tests/test_network.py
import numpy as np
import pytest

from plotomics import Network


def test_network_builds_traits_from_dicts():
    nodes = {
        "id": ["a", "b", "c"],
        "x": [0.0, 1.0, 2.0],
        "y": [0.0, 1.0, 0.0],
        "size": [4.0, 8.0, 6.0],
        "group": ["G1", "G1", "G2"],
        "label": ["Alpha", "Beta", "Gamma"],
    }
    edges = {"source": ["a", "b"], "target": ["b", "c"], "weight": [1.5, 2.0]}
    w = Network(
        nodes, edges, layout="precomputed", iterations=100, label_threshold=5
    )

    assert w.data["columns"]["id"] == ["a", "b", "c"]
    assert w.data["columns"]["source"] == ["a", "b"]
    assert w.data["columns"]["target"] == ["b", "c"]
    assert w.data["meta"]["nodeGroup"] == ["G1", "G1", "G2"]
    assert w.data["meta"]["nodeLabels"] == ["Alpha", "Beta", "Gamma"]
    assert w.options["layout"] == "precomputed"
    assert w.options["iterations"] == 100
    assert w.options["labelThreshold"] == 5

    # x, y, size (nodes) + weight (edges) packed as float32 columns.
    names = {c["name"] for c in w.schema["columns"]}
    assert names == {"x", "y", "size", "weight"}
    # 3 node coords/sizes (3 cols x 3) + 2 weights, all float32 (4 bytes).
    assert len(w.buffer) == (3 * 3 + 2) * 4


def test_network_omits_optional_columns():
    nodes = {"id": ["a", "b"]}
    edges = {"source": ["a"], "target": ["b"]}
    w = Network(nodes, edges)

    assert "nodeGroup" not in w.data["meta"]
    assert "nodeLabels" not in w.data["meta"]
    assert w.schema["columns"] == []
    assert w.buffer == b""
    assert w.options["layout"] == "forceatlas2"


def test_network_forwards_palette():
    w = Network(
        {"id": ["a", "b"]},
        {"source": ["a"], "target": ["b"]},
        palette=["#111111", "#222222"],
    )
    assert w.options["palette"] == ["#111111", "#222222"]


def test_network_requires_id():
    with pytest.raises(ValueError, match="`id`"):
        Network({"x": [1, 2]}, {"source": ["a"], "target": ["b"]})


def test_network_requires_source_and_target():
    with pytest.raises(ValueError, match="source` and `target`"):
        Network({"id": ["a", "b"]}, {"source": ["a"]})


def test_network_accepts_numpy_id_arrays():
    nodes = {"id": np.array(["a", "b"]), "size": np.array([3.0, 5.0])}
    edges = {"source": np.array(["a"]), "target": np.array(["b"])}
    w = Network(nodes, edges)
    assert w.data["columns"]["id"] == ["a", "b"]
    assert {c["name"] for c in w.schema["columns"]} == {"size"}


def test_network_rejects_duplicate_node_id():
    with pytest.raises(ValueError, match="duplicate node id"):
        Network({"id": ["a", "a"]}, {"source": ["a"], "target": ["a"]})


def test_network_rejects_dangling_edge():
    with pytest.raises(ValueError, match="not found among node ids"):
        Network({"id": ["a", "b"]}, {"source": ["a"], "target": ["z"]})


def test_network_precomputed_requires_xy():
    with pytest.raises(ValueError, match="precomputed"):
        Network(
            {"id": ["a", "b"]},
            {"source": ["a"], "target": ["b"]},
            layout="precomputed",
        )


def test_network_rejects_empty_nodes():
    with pytest.raises(ValueError, match="at least one row"):
        Network({"id": []}, {"source": [], "target": []})


def test_network_theme_in_options():
    w = Network(
        {"id": ["a", "b"]},
        {"source": ["a"], "target": ["b"]},
        theme={"background": "#222"},
    )
    assert w.options["theme"] == {"background": "#222"}


def test_network_forwards_directed():
    w = Network({"id": ["a", "b"]}, {"source": ["a"], "target": ["b"]})
    assert w.options["directed"] is False
    w2 = Network(
        {"id": ["a", "b"]}, {"source": ["a"], "target": ["b"]}, directed=True
    )
    assert w2.options["directed"] is True


def test_network_per_edge_color_column():
    w = Network(
        {"id": ["a", "b", "c"]},
        {"source": ["a", "b"], "target": ["b", "c"], "color": ["#f00", "#0f0"]},
    )
    assert w.data["columns"]["color"] == ["#f00", "#0f0"]

    w2 = Network({"id": ["a", "b"]}, {"source": ["a"], "target": ["b"]})
    assert "color" not in w2.data["columns"]


def test_network_rejects_mismatched_color_length():
    with pytest.raises(ValueError, match="`color` must match"):
        Network(
            {"id": ["a", "b", "c"]},
            {"source": ["a", "b"], "target": ["b", "c"], "color": ["#f00"]},
        )


def test_network_selected_trait_accepts_node_id_and_none():
    # The `selected` trait now holds a clicked node id (str) or None, not only a
    # list of point indices, so a network node click can populate it.
    w = Network({"id": ["a", "b"]}, {"source": ["a"], "target": ["b"]})
    w.selected = "a"
    assert w.selected == "a"
    w.selected = None
    assert w.selected is None
